third_party/blink/renderer/modules/cobalt/h5vcc_experiments/h_5_vcc_experiments.cc

```diff
@@ -0,0 +1,166 @@
+// Copyright 2025 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "third_party/blink/renderer/modules/cobalt/h5vcc_experiments/h_5_vcc_experiments.h"
+
+#include <string>
+
+#include "base/values.h"
+#include "cobalt/browser/cobalt_experiment_names.h"
+#include "cobalt/browser/h5vcc_experiments/public/mojom/h5vcc_experiments.mojom-blink.h"
+#include "third_party/blink/renderer/bindings/core/v8/script_promise.h"
+#include "third_party/blink/renderer/bindings/core/v8/script_promise_resolver.h"
+#include "third_party/blink/renderer/bindings/core/v8/v8_union_boolean_long_string.h"
+#include "third_party/blink/renderer/bindings/modules/v8/v8_experiment_configuration.h"
+#include "third_party/blink/renderer/core/frame/local_dom_window.h"
+
+namespace blink {
+
+H5vccExperiments::H5vccExperiments(LocalDOMWindow& window)
+    : ExecutionContextLifecycleObserver(window.GetExecutionContext()),
+      remote_h5vcc_experiments_(window.GetExecutionContext()) {}
+
+void H5vccExperiments::ContextDestroyed() {}
+
+ScriptPromise H5vccExperiments::setExperimentState(
+    ScriptState* script_state,
+    const ExperimentConfiguration* experiment_configuration,
+    ExceptionState& exception_state) {
+  auto* resolver = MakeGarbageCollected<ScriptPromiseResolver>(
+      script_state, exception_state.GetContext());
+  auto promise = resolver->Promise();
+
+  EnsureReceiverIsBound();
+
+  base::Value::Dict experiment_config_dict;
+
+  base::Value::Dict features;
+  for (auto& feature_name_and_value : experiment_configuration->features()) {
+    features.Set(feature_name_and_value.first.Utf8(),
+                 feature_name_and_value.second);
+  }
+  experiment_config_dict.Set(cobalt::kExperimentConfigFeatures,
+                             std::move(features));
+
+  base::Value::Dict feature_params;
+  std::string param_value;
+  for (auto& param_name_and_value : experiment_configuration->featureParams()) {
+    if (param_name_and_value.second->IsString()) {
+      param_value = param_name_and_value.second->GetAsString().Utf8();
+    } else if (param_name_and_value.second->IsLong()) {
+      param_value = std::to_string(param_name_and_value.second->GetAsLong());
+    } else if (param_name_and_value.second->GetAsBoolean()) {
+      param_value = "true";
+    } else if (!param_name_and_value.second->GetAsBoolean()) {
+      param_value = "false";
+    } else {
+      resolver->Reject();
```

Comment:
Add return here? Otherwise `resolver` will be reused in l.85 and a D/CHECK will be hit when it tries to be `Resolve()`d in `OnSetExperimentState()`. Maybe what we could do here is sth like:
```
    exception_state.ThrowDOMException(
        DOMExceptionCode::kInvalidStateError,  // Or something else, API defined.
        "Something went really really bad.");
    return;
```